Minor code improvements - WIP

cyberbono3 opened the pull request.

cyberbono3 committed:
WIP

uprendis commented on gossip/emitter/control.go:
```diff
@@ -62,6 +76,7 @@ func (em *Emitter) isAllowedToEmit(e inter.EventI, eTxs bool, metric ancestor.Me
 	adjustedPassedIdleTime := time.Duration(ancestor.Metric(passedTimeIdle/piecefunc.DecimalUnit) * metric)
 	passedBlocks := em.world.GetLatestBlockIndex() - em.prevEmittedAtBlock
 	// Forbid emitting if not enough power and power is decreasing
+	// Question: what is the intetion to declare code blocks with {...threshold = } to keep variables only within code block?
```
Yeah, just to separate the variables. Also visually separate the code

uprendis commented on gossip/emitter/emitter.go:
```diff
@@ -127,13 +128,11 @@ func (em *Emitter) init() {
 // Start starts event emission.
 func (em *Emitter) Start() {
-	if em.config.Validator.ID == 0 {
+	if em.config.Validator.ID == 0 || em.done != nil  {
```
Will need to update the comment to "short circuit if not a validator or already started"

uprendis commented on gossip/emitter/sync.go:
```diff
@@ -51,8 +51,7 @@ func (em *Emitter) currentSyncStatus() doublesign.SyncStatus {
 	}
-	prevEmitted := em.readLastEmittedEventID()
-	if prevEmitted != nil && (em.world.GetEvent(*prevEmitted) == nil && em.epoch <= prevEmitted.Epoch()) {
+	if prevEmitted := em.readLastEmittedEventID(); prevEmitted != nil && (em.world.GetEvent(*prevEmitted) == nil && em.epoch <= prevEmitted.Epoch()) {
```
Hmm, this one became too long

cyberbono3 replied:
Does multiline if statement sound good to you?  It does not make sense to declare `prevEmitted` as local in `currentSyncStatus` since it is used only inside if statement.
```
if prevEmitted := em.readLastEmittedEventID(); prevEmitted != nil &&
		em.world.GetEvent(*prevEmitted) == nil &&
		em.epoch <= prevEmitted.Epoch() {
		s.P2PSynced = time.Time{}
	}
```

uprendis replied:
```It does not make sense to declare prevEmitted as local```
Makes sense! The last variant looks good

uprendis commented on gossip/emitter/txs.go:
```diff
@@ -157,11 +156,10 @@ func (em *Emitter) addTxs(e *inter.MutableEventPayload, sorted *types.Transactio
 		sender, _ := types.Sender(em.world.TxSigner, tx)
 		// check transaction epoch rules
-		if epochcheck.CheckTxs(types.Transactions{tx}, rules) != nil {
+		if epochcheck.CheckTxs(types.Transactions{tx}, em.world.GetRules()) != nil {
```
nitpick: it might be less optimal to call the function in loop

uprendis commented on gossip/c_event_callbacks.go:
```diff
@@ -196,12 +196,15 @@ func (s *Service) processEvent(e *inter.EventPayload) error {
 	}
 
 	newEpoch := s.store.GetEpoch()
+	newEpochNotEqualOld := newEpoch != oldEpoch
```
You can name "sealed" or "epochSealed". "Epoch is sealed" means that it had ended and a next epoch has started

uprendis commented on gossip/emitter/control.go:
```diff
@@ -29,6 +29,20 @@ func updMetric(median, cur, upd idx.Event, validatorIdx idx.Validator, validator
+	and metric will be increased twice, is that expected behavior?
+
+	*/
+	// what will happen if e.g. seq = 1, metric = 0
```
We'll return 0.3

uprendis commented on gossip/emitter/control.go:
```diff
@@ -29,6 +29,20 @@ func updMetric(median, cur, upd idx.Event, validatorIdx idx.Validator, validator
 func eventMetric(orig ancestor.Metric, seq idx.Event) ancestor.Metric {
 	metric := ancestor.Metric(eventMetricF(uint64(orig)))
 	// kick start metric in a beginning of epoch, when there's nothing to observe yet
+	/* Question: are 2 if conditions below mutual exclusive or not? if they are mutual exclusive, switch statement can be a good option
```
Not exclusive

uprendis commented on gossip/emitter/emitter.go:
```diff
@@ -232,15 +233,12 @@ func (em *Emitter) getSortedTxs() *types.TransactionsByPriceAndNonce {
 	}
 	sortedTxs := em.getSortedTxs()
 
-	if em.world.IsBusy() {
```
It may make sense to leave the check closer to `em.world.Lock()` to reduce the chance of locking the mutex when there's an active block processing. The reason why we bother with `IsBusy` check is because `em.world.Lock` will lock engine mutex and then wait until block processing is complete (if any). That will make engine mutex unavailable while we're waiting for completion of block processing (in a case if it's running)

cyberbono3 replied:
I get your point. Reverting my changes.

uprendis commented on gossip/emitter/prev_action_files.go:
```diff
@@ -82,24 +71,35 @@ func (em *Emitter) writeLastEmittedEpochVote(e idx.Epoch) {
+	}
+
+	buf := make([]byte, bufLen)
+	if _, err := file.ReadAt(buf, 0); err != nil {
```
Technically we had/have a bug here, because ReadAt is not obliged to read the entire buffer. Practically it works only because buffer is very small

uprendis replied:
We may use `Seek` and then call `ioread.ReadAll` to read entire buffer. Or, we can wrap it with io.NewSectionReader (so that we wouldn't need to call `Seek`) and then call `ioread.ReadAll`

uprendis replied:
Oh sorry, I'm wrong:
```
// When ReadAt returns n < len(p), it returns a non-nil error
// explaining why more bytes were not returned. In this respect,
// ReadAt is stricter than Read.
```

cyberbono3 committed:
apply comments

cyberbono3 committed:
revert eventMetric

cyberbono3 committed:
WIP

cyberbono3 committed:
revert emitter change

cyberbono3 commented:
@uprendis  I have addressed your comments. I will continue working on it as soon as I have compete tasks with higher priority. Let us leave in draft for now.
